File: Common/src/main/java/com/cerbon/talk_balloons/client/BalloonRenderer.java
package com.cerbon.talk_balloons.client;

import com.cerbon.cerbons_api.api.general.data.HistoricalData;
import com.cerbon.talk_balloons.TalkBalloons;
import com.cerbon.talk_balloons.mixin.accessor.GuiGraphicsAccessor;
import com.cerbon.talk_balloons.util.TBConstants;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import it.unimi.dsi.fastutil.ints.IntArrayList;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.entity.EntityRenderDispatcher;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import org.joml.Matrix4f;
import org.joml.Quaternionf;
import org.joml.Vector3f;

import java.util.List;

@Environment(EnvType.CLIENT)
public class BalloonRenderer {
    private static final ResourceLocation BALLOON_TEXTURE = new ResourceLocation(TBConstants.MOD_ID, "textures/gui/balloon.png");

    private static final int MIN_BALLOON_WIDTH = TalkBalloons.config.minBalloonWidth;
    private static final int MAX_BALLOON_WIDTH = TalkBalloons.config.maxBalloonWidth;

    public static void renderBalloons(PoseStack poseStack, MultiBufferSource buffer, EntityRenderDispatcher entityRenderDispatcher, Font font, HistoricalData<String> messages, float playerHeight, int packedLight) {
        List<Integer> previousHeights = new IntArrayList();

        for (int i = 0; i < messages.size(); i++) {
            poseStack.pushPose();

            poseStack.translate(0.0, playerHeight + TalkBalloons.config.balloonsHeightOffset, 0.0D);
            poseStack.mulPose(Axis.YP.rotationDegrees(toEulerXyzDegrees(entityRenderDispatcher.cameraOrientation()).y));
            poseStack.scale(-0.025F, -0.025F, 0.025F);

            RenderSystem.enableDepthTest();
            RenderSystem.enablePolygonOffset();
            RenderSystem.polygonOffset(3.0F, 3.0F);

            Minecraft client = Minecraft.getInstance();
            GuiGraphics guiGraphics = GuiGraphicsAccessor.getGuiGraphics(client, poseStack, client.renderBuffers().bufferSource());

            String message = messages.get(i);
            int messageWidth = font.width(message);

            int balloonWidth = Mth.clamp(messageWidth, MIN_BALLOON_WIDTH, MAX_BALLOON_WIDTH);
            int balloonHeight = messageWidth > MAX_BALLOON_WIDTH ? Mth.ceil((float) messageWidth / MAX_BALLOON_WIDTH) : 1;

            if (balloonWidth % 2 == 0) // Width should be odd to correctly center the arrow
                balloonWidth--;

            int balloonsDistance = 0;
            for (int height : previousHeights)
                balloonsDistance += 9 * height + TalkBalloons.config.distanceBetweenBalloons;

            previousHeights.add(balloonHeight);

            int j = balloonHeight - 1;
            
            int baseX = balloonWidth / 2;
            int baseY = (-balloonHeight - j * 7) - j;

            // Left
            guiGraphics.blit(BALLOON_TEXTURE, -baseX - 2, baseY - balloonsDistance, 5, 5, 0.0F, 0.0F, 5, 5, 32, 32); // TOP
            guiGraphics.blit(BALLOON_TEXTURE, -baseX - 2, baseY + 5 - balloonsDistance, 5, balloonHeight + j * 8, 0.0F, 6.0F, 5, 1, 32, 32); // MID
            guiGraphics.blit(BALLOON_TEXTURE, -baseX - 2, 5 - balloonsDistance, 5, 5, 0.0F, 8.0F, 5, 5, 32, 32); // BOTTOM

            // Mid
            guiGraphics.blit(BALLOON_TEXTURE, -baseX + 3, baseY - balloonsDistance, balloonWidth - 4, 5, 6.0F, 0.0F, 5, 5, 32, 32); // TOP
            guiGraphics.blit(BALLOON_TEXTURE, -baseX + 3, baseY + 5 - balloonsDistance, balloonWidth - 4, balloonHeight + j * 8, 6.0F, 6.0F, 5, 1, 32, 32); // MID
            guiGraphics.blit(BALLOON_TEXTURE, -baseX + 3, 5 - balloonsDistance, balloonWidth - 4, 5, 6.0F, 8.0F, 5, 5, 32, 32); // BOTTOM

            // Right
            guiGraphics.blit(BALLOON_TEXTURE,  baseX - 2, baseY - balloonsDistance, 5, 5, 12.0F, 0.0F, 5, 5, 32, 32); // TOP
            guiGraphics.blit(BALLOON_TEXTURE,  baseX - 2, baseY + 5 - balloonsDistance, 5, balloonHeight + j * 8, 12.0F, 6.0F, 5, 1, 32, 32); // MID
            guiGraphics.blit(BALLOON_TEXTURE,  baseX - 2, 5 - balloonsDistance, 5, 5, 12.0F, 8.0F, 5, 5, 32, 32); // BOTTOM

            // Arrow
            RenderSystem.polygonOffset(2.0F, 2.0F);
            guiGraphics.blit(BALLOON_TEXTURE, -3, 9, 18, 6, 7, 4, 32, 32);

            RenderSystem.polygonOffset(0.0F, 0.0F);
            RenderSystem.disablePolygonOffset();

            Matrix4f matrix4f = poseStack.last().pose();
            font.drawInBatch(message, -messageWidth / 2.0F + 1, balloonHeight / 2.0F - balloonsDistance, TalkBalloons.config.textColor, false, matrix4f, buffer, Font.DisplayMode.NORMAL, 0, packedLight);

            poseStack.popPose();
        }
    }

    private static Vector3f toEulerXyz(Quaternionf quaternionf) {
        float f = quaternionf.w() * quaternionf.w();
        float g = quaternionf.x() * quaternionf.x();
        float h = quaternionf.y() * quaternionf.y();
        float i = quaternionf.z() * quaternionf.z();
        float j = f + g + h + i;
        float k = 2.0f * quaternionf.w() * quaternionf.x() - 2.0f * quaternionf.y() * quaternionf.z();
        float l = (float) Math.asin(k / j);

        if (Math.abs(k) > 0.999f * j)
            return new Vector3f(l, 2.0f * (float) Math.atan2(quaternionf.y(), quaternionf.w()), 0.0f);

        return new Vector3f(l, (float) Math.atan2(2.0f * quaternionf.x() * quaternionf.z() + 2.0f * quaternionf.y() * quaternionf.w(), f - g - h + i), (float) Math.atan2(2.0f * quaternionf.x() * quaternionf.y() + 2.0f * quaternionf.w() * quaternionf.z(), f - g + h - i));
    }

    private static Vector3f toEulerXyzDegrees(Quaternionf quaternionf) {
        Vector3f vec3f = BalloonRenderer.toEulerXyz(quaternionf);
        return new Vector3f((float) Math.toDegrees(vec3f.x()), (float) Math.toDegrees(vec3f.y()), (float) Math.toDegrees(vec3f.z()));
    }
}

File: common/src/main/kotlin/com/cerbon/talk_balloons/client/BalloonRenderer.kt
package com.cerbon.talk_balloons.client

import com.cerbon.talk_balloons.TalkBalloons
import com.cerbon.talk_balloons.client.resources.BalloonSpriteManager
import com.cerbon.talk_balloons.client.resources.BalloonStyle
import com.cerbon.talk_balloons.client.resources.BalloonStyleManager
import com.cerbon.talk_balloons.compat.CompatHandler
import com.cerbon.talk_balloons.compat.iris.IrisCompat
import com.cerbon.talk_balloons.util.HistoricalData
import com.cerbon.talk_balloons.util.SynchronizedConfigData
import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.BufferUploader
import com.mojang.blaze3d.vertex.DefaultVertexFormat
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.Tesselator
import com.mojang.blaze3d.vertex.VertexConsumer
import com.mojang.blaze3d.vertex.VertexFormat
import com.mojang.math.Axis
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.client.renderer.LightTexture
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.resources.metadata.gui.GuiSpriteScaling
import net.minecraft.network.chat.Component
import net.minecraft.util.FormattedCharSequence
import net.minecraft.util.Mth
import kotlin.math.min

object BalloonRenderer {
    @JvmField val SPRITE_MANAGER = BalloonSpriteManager(Minecraft.getInstance().textureManager)

    @JvmStatic
    fun renderBalloons(poseStack: PoseStack, cameraYaw: Float, font: Font, messages: HistoricalData<Component>, playerHeight: Float, configData: SynchronizedConfigData, light: Int) {
        if (messages.isEmpty())
            return

        val style = BalloonStyleManager.getStyleById(configData.balloonStyle)
        val balloonSprite = SPRITE_MANAGER.getSpriteAccess(style.balloon)
        val arrowSprite = SPRITE_MANAGER.getSpriteAccess(style.arrow)

        val consumer = Tesselator.getInstance().begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.PARTICLE)

        val padding = configData.balloonPadding
        val textColor = configData.textColor or (0xFF shl 24)
        val balloonTint = if (style.allowsTint)
            configData.balloonTint or (0xFF shl 24)
        else -1
        val fontHeight = font.lineHeight

        var balloonDistance = 0f

        messages.asReversed().forEachIndexed { index, message ->
            poseStack.pushPose()

            poseStack.translate(0.0f, playerHeight + TalkBalloons.config.balloonsHeightOffset, 0.0f)
            poseStack.mulPose(Axis.YP.rotationDegrees(-cameraYaw))
            poseStack.scale(-0.025f, -0.025f, -0.025f)

            val dividedMessage = font.split(message, TalkBalloons.config.maxBalloonWidth)
            val greatestTextWidth = dividedMessage.maxOf { font.width(it) }

            var textDistance = 0

            val balloonWidth = Mth.clamp(greatestTextWidth, TalkBalloons.config.minBalloonWidth, TalkBalloons.config.maxBalloonWidth)
            val actualBalloonWidth = balloonWidth + (padding * 2) + style.margins.horizontalMargins
            val baseX = -(actualBalloonWidth / 2f)

            if (dividedMessage.size > 1) {
                for (text in dividedMessage) {
                    drawString(poseStack.last(), font, text, -font.width(text) / 2f + 0.5f, -(fontHeight * dividedMessage.size) - balloonDistance + textDistance, textColor, false, light)
                    textDistance += fontHeight
                }
            } else {
                drawString(poseStack.last(), font, message.visualOrderText, -greatestTextWidth / 2f + 0.5f, -(fontHeight * dividedMessage.size) - balloonDistance, textColor, false, light)
                textDistance += fontHeight
            }

            val balloonHeight = textDistance + (padding * 2) + style.margins.verticalMargins - 2
            blitSprite(poseStack.last(), consumer, balloonSprite, baseX, -balloonDistance - balloonHeight, actualBalloonWidth, balloonHeight, balloonTint, light = light)

            if (index == 0) {
                blitSprite(poseStack.last(), consumer, arrowSprite, -(arrowSprite.contents().width() / 2f), -1f, arrowSprite.contents().width(), arrowSprite.contents().height(), balloonTint, 0.001f, light)
            }

            balloonDistance += balloonHeight + TalkBalloons.config.distanceBetweenBalloons

            poseStack.popPose()
        }

        val meshData = consumer.build()
        if (meshData != null) {
            RenderSystem.enableBlend()
            RenderSystem.defaultBlendFunc()
            RenderSystem.enableDepthTest()
            RenderSystem.enablePolygonOffset()
            RenderSystem.polygonOffset(3f, 3f)

            RenderSystem.setShader(GameRenderer::getParticleShader)
            RenderSystem.setShaderTexture(0, BalloonStyle.BALLOONS_SHEET)
            Minecraft.getInstance().gameRenderer.lightTexture().turnOnLightLayer()
            BufferUploader.drawWithShader(meshData)

            RenderSystem.disableDepthTest()
            RenderSystem.disableBlend()
            RenderSystem.disablePolygonOffset()
        }
    }

    private fun drawString(pose: PoseStack.Pose, font: Font, text: FormattedCharSequence, x: Float, y: Float, color: Int, dropShadow: Boolean, light: Int) {
        // otherwise text looks wonk af
        if (CompatHandler.isIrisLoaded && IrisCompat.isInShadowPass())
            return

        font.drawInBatch(text, x, y, color, dropShadow, pose.pose(), Minecraft.getInstance().renderBuffers().bufferSource(), Font.DisplayMode.NORMAL, 0, light)
    }

    private fun blitSprite(pose: PoseStack.Pose, consumer: VertexConsumer, sprite: TextureAtlasSprite, x: Float, y: Float, width: Int, height: Int, color: Int = -1, z: Float = 0f, light: Int) {
        val scaling = SPRITE_MANAGER.getMetadata(sprite).scaling

        if (scaling is GuiSpriteScaling.Stretch) {
            this.blitDirect(pose, consumer, x, y, width.toFloat(), height.toFloat(), sprite.u0, sprite.v0, sprite.u1, sprite.v1, color, z, light)
        } else if (scaling is GuiSpriteScaling.Tile) {
            this.blitTiled(pose, consumer, sprite, x, y, width, height, 0f, 0f, scaling.width, scaling.height, scaling.width, scaling.height, color, z, light)
        } else if (scaling is GuiSpriteScaling.NineSlice) {
            val border = scaling.border
            val leftBorder = border.left.coerceAtMost(width / 2)
            val rightBorder = border.right.coerceAtMost(width / 2)
            val topBorder = border.top.coerceAtMost(height / 2)
            val bottomBorder = border.bottom.coerceAtMost(height / 2)

            if (width == scaling.width && height == scaling.height) {
                this.blitFromSprite(pose, consumer, sprite, x, y, scaling.width, scaling.height, 0f, 0f, width.toFloat(), height.toFloat(), color, z, light)
            } else if (height == scaling.height()) {
                this.blitFromSprite(pose, consumer, sprite, x, y, scaling.width(), scaling.height(), 0f, 0.toFloat(), leftBorder.toFloat(), height.toFloat(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x + leftBorder, y, width - rightBorder - leftBorder, height, leftBorder.toFloat(), 0.toFloat(), scaling.width() - rightBorder - leftBorder, scaling.height(), scaling.width(), scaling.height(), color, z, light)
                this.blitFromSprite(pose, consumer, sprite, x + width - rightBorder, y, scaling.width(), scaling.height(), scaling.width() - rightBorder.toFloat(), 0.toFloat(), rightBorder.toFloat(), height.toFloat(), color, z, light)
            } else if (width == scaling.width()) {
                this.blitFromSprite(pose, consumer, sprite, x, y, scaling.width(), scaling.height(), 0.toFloat(), 0.toFloat(), width.toFloat(), topBorder.toFloat(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x, y + topBorder, width, height - bottomBorder - topBorder, 0.toFloat(), topBorder.toFloat(), scaling.width(), scaling.height() - bottomBorder - topBorder, scaling.width(), scaling.height(), color, z, light)
                this.blitFromSprite(pose, consumer, sprite, x, y + height - bottomBorder, scaling.width(), scaling.height(), 0.toFloat(), scaling.height() - bottomBorder.toFloat(), width.toFloat(), bottomBorder.toFloat(), color, z, light)
            } else {
                this.blitFromSprite(pose, consumer, sprite, x, y, scaling.width(), scaling.height(), 0.toFloat(), 0.toFloat(), leftBorder.toFloat(), topBorder.toFloat(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x + leftBorder, y, width - rightBorder - leftBorder, topBorder, leftBorder.toFloat(), 0f, scaling.width() - rightBorder - leftBorder, topBorder, scaling.width(), scaling.height(), color, z, light)
                this.blitFromSprite(pose, consumer, sprite, x + width - rightBorder, y, scaling.width(), scaling.height(), scaling.width() - rightBorder.toFloat(), 0f, rightBorder.toFloat(), topBorder.toFloat(), color, z, light)

                this.blitFromSprite(pose, consumer, sprite, x, y + height - bottomBorder, scaling.width(), scaling.height(), 0f, scaling.height() - bottomBorder.toFloat(), leftBorder.toFloat(), bottomBorder.toFloat(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x + leftBorder, y + height - bottomBorder, width - rightBorder - leftBorder, bottomBorder, leftBorder.toFloat(), scaling.height() - bottomBorder.toFloat(), scaling.width() - rightBorder - leftBorder, bottomBorder, scaling.width(), scaling.height(), color, z, light)
                this.blitFromSprite(pose, consumer, sprite, x + width - rightBorder, y + height - bottomBorder, scaling.width(), scaling.height(), scaling.width() - rightBorder.toFloat(), scaling.height() - bottomBorder.toFloat(), rightBorder.toFloat(), bottomBorder.toFloat(), color, z, light)

                this.blitTiled(pose, consumer, sprite, x, y + topBorder, leftBorder, height - bottomBorder - topBorder, 0f, topBorder.toFloat(), leftBorder, scaling.height() - bottomBorder - topBorder, scaling.width(), scaling.height(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x + leftBorder, y + topBorder, width - rightBorder - leftBorder, height - bottomBorder - topBorder, leftBorder.toFloat(), topBorder.toFloat(), scaling.width() - rightBorder - leftBorder, scaling.height() - bottomBorder - topBorder, scaling.width(), scaling.height(), color, z, light)
                this.blitTiled(pose, consumer, sprite, x + width - rightBorder, y + topBorder, leftBorder, height - bottomBorder - topBorder, scaling.width() - rightBorder.toFloat(), topBorder.toFloat(), rightBorder, scaling.height() - bottomBorder - topBorder, scaling.width(), scaling.height(), color, z, light)
            }
        }
    }

    private fun blitTiled(pose: PoseStack.Pose, consumer: VertexConsumer, sprite: TextureAtlasSprite, x: Float, y: Float, width: Int, height: Int, uOffset: Float, vOffset: Float, uWidth: Int, vHeight: Int, nineSliceWidth: Int, nineSliceHeight: Int, color: Int = -1, z: Float = 0f, light: Int) {
        if (width > 0 && height > 0) {
            if (uWidth <= 0 || vHeight <= 0)
                throw IllegalArgumentException("Tiled sprite texture size must be positive, got ${uWidth}x${vHeight}")

            var i = 0
            while (i < width) {
                val j = min(uWidth, width - i)

                var k = 0
                while (k < height) {
                    val l = min(vHeight, height - k)
                    this.blitFromSprite(pose, consumer, sprite, x + i, y + k, nineSliceWidth, nineSliceHeight, uOffset, vOffset, j.toFloat(), l.toFloat(), color, z, light)
                    k += vHeight
                }

                i += uWidth
            }
        }
    }

    private fun blitFromSprite(pose: PoseStack.Pose, consumer: VertexConsumer, sprite: TextureAtlasSprite, x: Float, y: Float, texWidth: Int, texHeight: Int, uOffset: Float, vOffset: Float, uWidth: Float, vHeight: Float, color: Int = -1, z: Float = 0f, light: Int) {
        this.blitDirect(pose, consumer, x, y, uWidth, vHeight,
            sprite.getU(uOffset / texWidth), sprite.getV(vOffset / texHeight),
            sprite.getU((uOffset + uWidth) / texWidth), sprite.getV((vOffset + vHeight) / texHeight),
            color, z, light
        )
    }

    private fun blit(pose: PoseStack.Pose, consumer: VertexConsumer, x: Float, y: Float, width: Int, height: Int, uOffset: Float, vOffset: Float, uWidth: Float, vHeight: Float, color: Int = -1, z: Float = 0f, light: Int) {
        this.blitDirect(pose, consumer, x, y, width.toFloat(), height.toFloat(), uOffset, vOffset, uOffset + uWidth, vOffset + vHeight, color, z, light)
    }

    private fun blitDirect(pose: PoseStack.Pose, consumer: VertexConsumer, x: Float, y: Float, width: Float, height: Float, u0: Float, v0: Float, u1: Float, v1: Float, color: Int = -1, z: Float = 0f, light: Int) {
        val x2 = (x + width)
        val y2 = (y + height)

        consumer.addVertex(pose, x, y, z)
            .setUv(u0, v0)
            .setColor(color)
            .setLight(light)

        consumer.addVertex(pose, x, y2, z)
            .setUv(u0, v1)
            .setColor(color)
            .setLight(light)

        consumer.addVertex(pose, x2, y2, z)
            .setUv(u1, v1)
            .setColor(color)
            .setLight(light)

        consumer.addVertex(pose, x2, y, z)
            .setUv(u1, v0)
            .setColor(color)
            .setLight(light)
    }
}
